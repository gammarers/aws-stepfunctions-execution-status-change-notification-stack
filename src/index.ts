import { StepFunctionsExecutionStatusChangeDetectionEventRule } from '@gammarers/aws-stepfunctions-execution-status-change-detection-event-rule';
import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';
import { NotificationStateMachine } from './resources/notification-state-machine';

export interface Notifications {
  readonly emails?: string[];
  // readonly slack?: Slack;
}

export interface StepFunctionsExecutionStatueChangeNotificationStackProps extends StackProps {
  // readonly targetResource: TargetResource;
  readonly enableRule?: boolean;
  readonly notifications?: Notifications;
  // readonly resourceNamingOption?: ResourceNamingOption;
}

export class StepFunctionsExecutionStatueChangeNotificationStack extends Stack {
  constructor(scope: Construct, id: string, props: StepFunctionsExecutionStatueChangeNotificationStackProps) {
    super(scope, id, props);

    // ðŸ‘‡ SNS Topic for notifications
    const topic: sns.Topic = new sns.Topic(this, 'NotificationTopic', {
      // topicName: names.notificationTopicName,
      // displayName: names.notificationTopicDisplayName,
    });

    // ðŸ‘‡ Subscribe an email endpoint to the topic
    const emails = props.notifications?.emails ?? [];
    for (const email of emails) {
      topic.addSubscription(new subscriptions.EmailSubscription(email));
    }

    // ðŸ‘‡ Create State Machine
    const stateMachine = new NotificationStateMachine(this, 'StateMachine', {
      stateMachineName: undefined,
      timeout: Duration.minutes(1),
      notificationTopic: topic,
    });

    // @ts-ignore
    new StepFunctionsExecutionStatusChangeDetectionEventRule(this, 'StepFunctionsExecutionStatusChangeDetectionEventRule', {
      targets: [
        new targets.SfnStateMachine(stateMachine, {
          input: events.RuleTargetInput.fromObject({
            event: events.EventField.fromPath('$'),
            params: {
              // tagKey: props.targetResource.tagKey,
              // tagValues: props.targetResource.tagValues,
            },
          }),
        }),
      ],
    });
  }
}