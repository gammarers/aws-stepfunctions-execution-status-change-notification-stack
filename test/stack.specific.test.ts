import { App, Duration } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StepFunctionsExecutionStatueChangeNotificationStack } from '../src';

describe('Stack Specific Testing', () => {

  const app = new App();

  const stack = new StepFunctionsExecutionStatueChangeNotificationStack(app, 'StepFunctionsExecutionStatueChangeNotificationStack', {
    enableRule: true,
    notifications: {
      emails: [
        'foo@example.com',
        'bar@example.net',
      ],
    },
    timeoutOption: {
      stateMachineTimeout: Duration.minutes(10),
    },
  });

  const template = Template.fromStack(stack);

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
