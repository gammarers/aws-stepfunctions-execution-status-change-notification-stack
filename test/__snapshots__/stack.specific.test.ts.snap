// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Specific Testing Should match snapshot 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "NotificationTopicEB7A0DF1": {
      "Type": "AWS::SNS::Topic",
    },
    "NotificationTopicbarexamplenetB2CEEB3B": {
      "Properties": {
        "Endpoint": "bar@example.net",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "NotificationTopicEB7A0DF1",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "NotificationTopicfooexamplecom623834B6": {
      "Properties": {
        "Endpoint": "foo@example.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "NotificationTopicEB7A0DF1",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"InitDefinition","States":{"InitDefinition":{"Type":"Pass","Result":{"Statues":[{"key":"RUNNING","value":"Running","emoji":"ü•≥"},{"key":"SUCCEEDED","value":"Succeeded","emoji":"ü§©"},{"key":"FAILED","value":"Failed","emoji":"üò´"},{"key":"TIMED_OUT","value":"TimeOut","emoji":"ü§î"},{"key":"ABORTED","value":"Aborted","emoji":"üòÆ"},{"key":"PENDING_REDRIVE","value":"PenfdingRedrive","emoji":"üßê"}]},"ResultPath":"$.Definition","Next":"isNotMe"},"isNotMe":{"Type":"Choice","Choices":[{"Not":{"Variable":"$$.StateMachine.Id","StringEqualsPath":"$.event.detail.stateMachineArn"},"Next":"PrepareTopicValue"}],"Default":"Skip"},"Skip":{"Type":"Pass","Next":"Succeed"},"Succeed":{"Type":"Succeed"},"PublishMessage":{"Next":"Succeed","Type":"Task","ResultPath":"$.Result.Sns","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish","Parameters":{"TopicArn":"",
              {
                "Ref": "NotificationTopicEB7A0DF1",
              },
              "","Message":{"default.$":"$.Generate.Topic.TextMessage","email.$":"$.Generate.Topic.TextMessage"},"MessageStructure":"json","Subject.$":"$.Generate.Topic.Subject"}},"GenerateTopicValue":{"Type":"Pass","ResultPath":"$.Generate.Topic","Parameters":{"Subject.$":"States.Format('{} [{}] AWS Step Functions Execution State Notification [{}][{}]', $.Prepare.Topic.Values.Emoji, $.Prepare.Topic.Values.Status, $.event.account, $.event.region)","TextMessage.$":"States.Format('Account : {}\\nRegion : {}\\nState Machine : {}\\nExecution Name : {}\\n\\n{}', $.event.account, $.event.region, $.Prepare.Topic.Values.StateMachieName, $.Prepare.Topic.Values.ExecutionName, States.Format('https://console.aws.amazon.com/states/home?{}#/v2/executions/details/{}', $.event.region, $.Prepare.Topic.Values.ExecutionArn))"},"Next":"PublishMessage"},"PrepareTopicValue":{"Type":"Pass","ResultPath":"$.Prepare.Topic.Values","Parameters":{"StateMachieName.$":"States.ArrayGetItem(States.StringSplit($.event.detail.stateMachineArn, ':'), 6)","ExecutionName.$":"$.event.detail.name","ExecutionArn.$":"$.event.detail.executionArn","Emoji.$":"States.ArrayGetItem($.Definition.Statues[?(@.key == $.event.detail.status)].emoji, 0)","Status.$":"States.ArrayGetItem($.Definition.Statues[?(@.key == $.event.detail.status)].value, 0)","ContextStateMachineId.$":"$$.StateMachine.Id"},"Next":"GenerateTopicValue"}},"TimeoutSeconds":600}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineEventsRoleDBCDECD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEventsRoleDefaultPolicyFB602CA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEventsRoleDefaultPolicyFB602CA9",
        "Roles": [
          {
            "Ref": "StateMachineEventsRoleDBCDECD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "NotificationTopicEB7A0DF1",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StepFunctionsExecutionStatusChangeDetectionEventRule270F3930": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "status": [
              "FAILED",
              "TIMED_OUT",
            ],
          },
          "detail-type": [
            "Step Functions Execution Status Change",
          ],
          "source": [
            "aws.states",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "f1": "$",
              },
              "InputTemplate": "{"event":<f1>,"params":{}}",
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "StateMachineEventsRoleDBCDECD1",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
